File: scripts/register-webhook.py
"""
scripts/register-webhook.py

Registers the Telegram Bot webhook with the deployed Vercel app URL.
Uses only Python stdlib — no packages required.
"""
import json
import os
import sys
import urllib.request
import urllib.error

BOT_TOKEN = os.environ.get("TELEGRAM_BOT_TOKEN", "8676026319:AAFmZ0kdiAPbMXLpsJJY6fN_uxZ78QxCN-0")
APP_URL   = os.environ.get("NEXT_PUBLIC_APP_URL", "https://v0-bidpilot-ai-saas.vercel.app").rstrip("/")
SECRET    = os.environ.get("TELEGRAM_WEBHOOK_SECRET", "IvanivAngelina15032008")

WEBHOOK_URL = f"{APP_URL}/api/webhook"
API_BASE    = f"https://api.telegram.org/bot{BOT_TOKEN}"

def tg(method, payload=None):
    url = f"{API_BASE}/{method}"
    data = json.dumps(payload).encode() if payload else None
    headers = {"Content-Type": "application/json"} if data else {}
    req = urllib.request.Request(url, data=data, headers=headers, method="POST" if data else "GET")
    try:
        with urllib.request.urlopen(req, timeout=15) as r:
            return json.loads(r.read().decode())
    except urllib.error.HTTPError as e:
        return json.loads(e.read().decode())
    except Exception as exc:
        return {"ok": False, "description": str(exc)}

def main():
    print("=" * 55)
    print("  BidPilot AI — Telegram Webhook Registration")
    print("=" * 55)
    print(f"Bot token (masked) : {BOT_TOKEN[:12]}...")
    print(f"App URL            : {APP_URL}")
    print(f"Webhook URL        : {WEBHOOK_URL}")
    print(f"Secret (masked)    : {SECRET[:4]}...")
    print()

    # 1. Verify bot token
    me = tg("getMe")
    if not me.get("ok"):
        print(f"ERROR: getMe failed — {me.get('description', 'invalid token?')}")
        sys.exit(1)
    bot = me["result"]
    print(f"Bot verified       : {bot['first_name']} (@{bot.get('username', 'N/A')})")
    print(f"Bot ID             : {bot['id']}")
    print()

    # 2. Set webhook
    result = tg("setWebhook", {
        "url": WEBHOOK_URL,
        "secret_token": SECRET,
        "allowed_updates": ["message", "callback_query", "inline_query"],
        "drop_pending_updates": True,
    })
    print("setWebhook response:")
    print(json.dumps(result, indent=2, ensure_ascii=False))
    print()

    if not result.get("ok"):
        print(f"ERROR: {result.get('description', 'unknown error')}")
        sys.exit(1)

    # 3. Verify registration
    info = tg("getWebhookInfo")
    wi = info.get("result", {})
    print("Webhook info:")
    print(f"  url                  : {wi.get('url', '(none)')}")
    print(f"  pending_update_count : {wi.get('pending_update_count', 0)}")
    print(f"  last_error_message   : {wi.get('last_error_message', '(none)')}")
    print(f"  max_connections      : {wi.get('max_connections', 'N/A')}")
    print()

    registered_url = wi.get("url", "")
    if registered_url == WEBHOOK_URL:
        print("SUCCESS: Webhook registered and verified.")
        print()
        print("Next step — open your bot in Telegram and send /start:")
        print(f"  https://t.me/{bot.get('username', 'your_bot')}")
    else:
        print(f"WARNING: registered URL does not match.")
        print(f"  expected : {WEBHOOK_URL}")
        print(f"  got      : {registered_url}")

if __name__ == "__main__":
    main()

File: scripts/register-webhook.mjs
/**
 * scripts/register-webhook.mjs
 *
 * Registers the Telegram Bot webhook with the deployed Vercel app URL.
 * Run after every new deployment:
 *
 *   node scripts/register-webhook.mjs
 *
 * Reads env vars from process.env (set via .env.local or shell).
 */

import { readFileSync } from 'fs';
import { resolve } from 'path';

// ── Load .env.local if running locally ───────────────────────────────────────
try {
  const envPath = resolve(process.cwd(), '.env.local');
  const raw = readFileSync(envPath, 'utf-8');
  for (const line of raw.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const eqIdx = trimmed.indexOf('=');
    if (eqIdx < 0) continue;
    const key = trimmed.slice(0, eqIdx).trim();
    const value = trimmed.slice(eqIdx + 1).trim().replace(/^['"]|['"]$/g, '');
    if (!process.env[key]) process.env[key] = value;
  }
  console.log('Loaded .env.local');
} catch {
  // No .env.local — rely on process.env (CI / Vercel)
}

// ── Validate required vars ────────────────────────────────────────────────────
const BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const APP_URL = (process.env.NEXT_PUBLIC_APP_URL ?? '').replace(/\/$/, '');
const SECRET = process.env.TELEGRAM_WEBHOOK_SECRET ?? '';

if (!BOT_TOKEN) {
  console.error('ERROR: TELEGRAM_BOT_TOKEN is not set');
  process.exit(1);
}
if (!APP_URL) {
  console.error('ERROR: NEXT_PUBLIC_APP_URL is not set');
  process.exit(1);
}

const WEBHOOK_URL = `${APP_URL}/api/webhook`;
const TELEGRAM_API = `https://api.telegram.org/bot${BOT_TOKEN}`;

// ── Register webhook ──────────────────────────────────────────────────────────
console.log(`\nRegistering webhook: ${WEBHOOK_URL}\n`);

const res = await fetch(`${TELEGRAM_API}/setWebhook`, {
  method: 'POST',
  headers: { 'Content-Type': 'application/json' },
  body: JSON.stringify({
    url: WEBHOOK_URL,
    secret_token: SECRET,
    allowed_updates: ['message', 'callback_query', 'inline_query'],
    drop_pending_updates: true,
  }),
});

const data = await res.json();
console.log('setWebhook result:', JSON.stringify(data, null, 2));

// ── Verify ────────────────────────────────────────────────────────────────────
const infoRes = await fetch(`${TELEGRAM_API}/getWebhookInfo`);
const info = await infoRes.json();
console.log('\ngetWebhookInfo:', JSON.stringify(info.result, null, 2));

// ── Bot info ──────────────────────────────────────────────────────────────────
const meRes = await fetch(`${TELEGRAM_API}/getMe`);
const me = await meRes.json();
console.log('\nBot:', me.result?.first_name, `(@${me.result?.username})`);

if (data.ok) {
  console.log('\nWebhook registered successfully.');
} else {
  console.error('\nFailed to register webhook:', data.description);
  process.exit(1);
}
